### PR TITLE
doc: add document for charge station based routing

### LINKED ISSUE #366: Add document for charge station based routing
subtask of #358

### PR DESCRIPTION
**Pull request by CodeBear801**
issue: https://github.com/Telenav/osrm-backend/issues/366

# Issue

- Targeting issue: the issue will be CLOSED once the PR merged. If there is no issue that addresses the problem, please open a corresponding issue and link it here. Uses the [Closes keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) to close them automatically.     
Closes #366 

- Any other related issue? Mention them here.    

## Description    
A summary description for what the PR achieves.     

Add document for charge station based routing      

## Tasklist

 - [x] CHANGELOG-FORK.md entry ([CHANGELOG](https://github.com/Telenav/osrm-backend/wiki/CHANGELOG))
 - [ ] [profiles/CHANGELOG.md](https://github.com/Telenav/osrm-backend/blob/master/profiles/CHANGELOG.md) entry if any `Lua` changes   
 - [ ] update relevant [Wiki pages](https://github.com/Project-OSRM/osrm-backend/wiki)
 - [ ] add tests (see [testing documentation](https://github.com/Project-OSRM/osrm-backend/blob/master/docs/testing.md)

## Prerequirements
- Want to contribute? Great! First, please read this page [Contribution Guidelines](https://github.com/Telenav/osrm-backend/wiki/Contribution-Guidelines).    
- If your PR is still work in progress please attach the relevant label.

### CODE REVIEW

File: integration/doc/oasis/charge_station_based_routing.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+
+## General idea
+
+If `Orig` and `Dest` has some distance in between, let's say needs at least one charge time.
+We could find all charge stations which could be reached for start point and all charge stations are needed to reach destination, if there are overlap in between, then we could end up with single charge station, otherwise, calculate the best route based on charge station network, node is charge station, edge is reachable charge station for each charge station.
+
+![](../graph/charge_station_based_routing.png)
+
+If we pro-process all charge stations in the graph and record the `connectivity` for each charge station, which means for each single charge station, we could retrieve  **all reachable charge stations sorted by distance or energy needed**, then we could calculate optimal result based on classic algorithm.
```
Comment (wangyoucao577) [inline integration/doc/oasis/charge_station_based_routing.md:16]: pro-process -> pre-process

File: integration/doc/oasis/charge_station_based_routing.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+### Query
+- Given two points of orig/dest
+- Check whether orig/dest is reachable by current energy
+- If not start two strategy together: `charge station based routing` and `search along route`
+- logic of `charge station routing`
+   + For input, need using its current energy to filter all charge station it could reached
+   + For Dest, check all charge station needed to reach destination
+        *  Check whether there is overlap between those two, if yes, then go to next step
+        * For all reachable stations from start, generate a reachable cellids by combine each charge station's 1-time-charge reachable cells result, see whether there is any overlap with dest's stations(or cellids contain those stations), if yes, go to next step
+        * For each reachable charge station of start, they have their unique value of `arrival energy`, this value need to be considered during calculating `charge times`, but the exist status of each charge station is fixed: such as charge to 60% of energy, 80% of energy, etc.
```
Comment (wangyoucao577) [inline integration/doc/oasis/charge_station_based_routing.md:71]: What does the meaning of `exist status`?

File: integration/doc/oasis/charge_station_based_routing.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  + [#236](https://github.com/Telenav/osrm-backend/issues/236)
+
+### Query
+- Given two points of orig/dest
+- Check whether orig/dest is reachable by current energy
+- If not start two strategy together: `charge station based routing` and `search along route`
+- logic of `charge station routing`
+   + For input, need using its current energy to filter all charge station it could reached
+   + For Dest, check all charge station needed to reach destination
+        *  Check whether there is overlap between those two, if yes, then go to next step
```
Comment (wangyoucao577) [inline integration/doc/oasis/charge_station_based_routing.md:69]: Question: if no overlap, currently the algorithm will be ignored, right? On the other hand, we at least already got some mandatory stations. Find shortest path on the station graph with given orig/dest and these marked mandatory stations may also possible, how's your opinion?

File: integration/doc/oasis/charge_station_based_routing.md (hunk 2 above)
Comment (CodeBear801) [reply integration/doc/oasis/charge_station_based_routing.md:71]: * For each reachable charge station of start, they have their unique value of `arrival energy`, this value need to be considered during calculating `charge times`, but the `exist status` of each charge station is fixed: such as charge to 60% of energy, 80% of energy, etc.

image

Comment (wangyoucao577) [reply integration/doc/oasis/charge_station_based_routing.md:71]: So it's `target charging energy` when charge on this station, right? The name `exist status` does not make sense, may be it's a typo that you want to use `exit status`?

Comment (CodeBear801) [reply integration/doc/oasis/charge_station_based_routing.md:71]: You are right, using `target energy` is better, which is more align with code

```go

// definition
// State contains charging related information
type State struct {
	Energy float64
}

// CreateChargingStates generates charging status
func (f *chargeStrategy) CreateChargingStates() []State {
}


// client
for _, targetState := range g.strategy.CreateChargingStates() {

}
```

File: integration/doc/oasis/charge_station_based_routing.md (hunk 3 above)
Comment (CodeBear801) [reply integration/doc/oasis/charge_station_based_routing.md:69]: It's possible.  I think the idea logic is:
- generate an iterator represent all reachable charge stations from orig, which represent `mandatory stations`
- generate an iterator represent all reachable charge stations from dest, which represent `mandatory stations`
- if no overlap, those iterators will be used in graph for next step

`Iterators` has been implemented but I didn't used it in graph calculation yet.

File: integration/doc/oasis/charge_station_based_routing.md (hunk 2 above)
Comment (CodeBear801) [reply integration/doc/oasis/charge_station_based_routing.md:71]: I changed the description to `target status` I think which makes more sense.

**Pull request merged**